pkg/cli/sanitize/hide_secrets.go

```diff
@@ -0,0 +1,207 @@
+	resources := strings.Split(manifest, "\n---")
+	outRes := make([]string, 0, len(resources))
+
+	for _, r := range resources {
+		var resourceMap map[string]interface{}
+		err := yaml.Unmarshal([]byte(r), &resourceMap)
+		if err != nil {
+			return "", errors.Wrapf(err, "failed to unmarshal %q resource", tryToGetName(resourceMap))
+		}
+
+		if isSecret(resourceMap) {
+			rs, err := hideSecretData(r)
+			if err != nil {
+				return "", errors.Wrapf(err, "failed to hide %q Secret data", tryToGetName(resourceMap))
+			}
+			r = rs
+		}
+
+		outRes = append(outRes, r)
+	}
+
+	return strings.Join(outRes, "\n---"), nil
+}
+
+func tryToGetName(resourceMap map[string]interface{}) string {
+	metadata, ok := resourceMap["metadata"].(map[string]interface{})
+	if !ok {
+		return ""
+	}
+
+	name, ok := metadata["name"].(string)
+	if !ok {
+		return ""
+	}
+
+	return name
+}
+
+func isSecret(resource map[string]interface{}) bool {
+	kind, ok := resource["kind"].(string)
+	if !ok || kind != "Secret" {
+		return false
+	}
+
+	apiVersion, ok := resource["apiVersion"].(string)
+	if !ok || apiVersion != "v1" {
+		return false
+	}
+
+	return true
+}
+
+func hideSecretData(raw string) (string, error) {
+	lines := strings.Split(raw, "\n")
+	outLines := make([]string, 0, len(lines))
+
+	// To minimize impact of empty lines and custom indentation
+	// we only marshal `data` and `secretData` sections of the Secrets
+	for i := 0; i < len(lines); i++ {
+		line := lines[i]
+		outLines = append(outLines, line)
+
+		if strings.HasPrefix(line, "data:") || strings.HasPrefix(line, "stringData:") {
+			endLine := findSectionEnd(lines, i+1)
+			sanitizedLines, err := sanitizeDataYaml(lines[i+1 : endLine+1])
+			if err != nil {
+				return "", errors.Wrap(err, "failed to sanitize Secret data")
+			}
+			outLines = append(outLines, sanitizedLines...)
+
+			i = endLine
+		}
+	}
+
+	return strings.Join(outLines, "\n"), nil
+}
+
+func sanitizeDataYaml(yamlLines []string) ([]string, error) {
+	yamlData := strings.Join(yamlLines, "\n")
+	data := &yaml.Node{}
+	err := yaml.Unmarshal([]byte(yamlData), data)
+	if err != nil {
+		return nil, err
+	}
+	if len(data.Content) == 0 {
+		return []string{}, nil
+	}
+
+	node := data.Content[0]
+	sanitizeNode(node)
+
+	// Try to preserve indentation of the data section
+	indent := "  "
+	if len(node.Content) > 1 {
+		lineNum := node.Content[1].Line
+		indent = takeWhitespace(yamlLines[lineNum-1])
+	}
+
+	sanitized, err := yaml.Marshal(node)
+	if err != nil {
+		return nil, err
+	}
+	str := strings.TrimSpace(string(sanitized))
+	lines := strings.Split(str, "\n")
+
+	for i := range lines {
+		lines[i] = fmt.Sprintf("%s%s", indent, lines[i])
+	}
+
+	return lines, nil
+}
+
+func sanitizeNode(node *yaml.Node) {
+	if node.Kind != yaml.MappingNode {
+		return
+	}
+	contents := node.Content
+
+	for i := 1; i < len(contents); i += 2 {
+		contents[i].Style = 0 // Erase literal style
+		contents[i].SetString(hiddenSecretValue)
```

Comment:
can you help me understand this? why hardcoded `i += 2` ?

Reply:
Sure, every other `*yaml.Node` represents the the key so it is not sanitised. 
For example:
```
key1: val1
key2: val2
```
`key1` will be Node 0, `val1` will be Node 1, `key2` will be Node 3 etc.

Reply:
All I could see in the test cases were string values.

Is there a test case for if null values are passed as the value of `node`? What about integral, boolean, or floating point numbers?

How does it handle nested values?

```
foo:
  bar: car
```

What about anchor tags?

```
foo: &bar
```

Without any validation, these are all acceptable input based on the function signature, which can bring up a number of security issues.

This entire `sanitizeDataYaml` function seems like really, really brittle code. I'm not sure I'm comfortable with maintaining this code as it currently stands.